File: src-tauri/src/pdf/numbering.rs
use crate::pdf::page_label::PageLabelNumberingStyle;

pub struct Numbering;

impl Numbering {
    pub fn to_roman(number: i32, upper_case: bool) -> String {
        if upper_case {
            Self::to_roman_upper_case(number)
        } else {
            Self::to_roman_lower_case(number)
        }
    }

    pub fn to_roman_lower_case(number: i32) -> String {
        Self::to_roman_upper_case(number).to_lowercase()
    }

    pub fn to_roman_upper_case(number: i32) -> String {
        if number < 1 || number > 3999 {
            return String::new();
        }
        let thousands = ["", "M", "MM", "MMM"];
        let hundreds = ["", "C", "CC", "CCC", "CD", "D", "DC", "DCC", "DCCC", "CM"];
        let tens = ["", "X", "XX", "XXX", "XL", "L", "LX", "LXX", "LXXX", "XC"];
        let ones = ["", "I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX"];

        let idx_thousands = (number / 1000) as usize;
        let idx_hundreds = ((number % 1000) / 100) as usize;
        let idx_tens = ((number % 100) / 10) as usize;
        let idx_ones = (number % 10) as usize;

        format!(
            "{}{}{}{}",
            thousands[idx_thousands], hundreds[idx_hundreds], tens[idx_tens], ones[idx_ones]
        )
    }

    pub fn to_latin_alphabet_number(number: i32, upper_case: bool) -> String {
        if upper_case {
            Self::to_latin_alphabet_number_upper_case(number)
        } else {
            Self::to_latin_alphabet_number_lower_case(number)
        }
    }

    pub fn to_latin_alphabet_number_lower_case(number: i32) -> String {
        Self::to_latin_alphabet_number_upper_case(number).to_lowercase()
    }

    pub fn to_latin_alphabet_number_upper_case(mut number: i32) -> String {
        if number <= 0 {
            return String::new();
        }
        let mut sb = String::new();
        while number > 0 {
            number -= 1;
            let char_code = (b'A' + (number % 26) as u8) as char;
            sb.insert(0, char_code);
            number /= 26;
        }
        sb
    }

    pub fn format_page_number(
        style: &PageLabelNumberingStyle,
        number: i32,
        prefix: Option<&str>,
    ) -> String {
        let label = match style {
            PageLabelNumberingStyle::None => String::new(),
            PageLabelNumberingStyle::DecimalArabicNumerals => number.to_string(),
            PageLabelNumberingStyle::UppercaseRomanNumerals => Self::to_roman_upper_case(number),
            PageLabelNumberingStyle::LowercaseRomanNumerals => Self::to_roman_lower_case(number),
            PageLabelNumberingStyle::UppercaseLetters => Self::to_latin_alphabet_number_upper_case(number),
            PageLabelNumberingStyle::LowercaseLetters => Self::to_latin_alphabet_number_lower_case(number),
        };

        if let Some(p) = prefix {
            format!("{}{}", p, label)
        } else {
            label
        }
    }
}

File: src/lib/pdf-processing/numbering.ts
import { PageLabelNumberingStyle } from '../types/page-label.ts';

export class Numbering {
    
    static toRoman(number: number, upperCase: boolean): string {
        return upperCase ? this.toRomanUpperCase(number) : this.toRomanLowerCase(number);
    }

    static toRomanLowerCase(number: number): string {
        return this.toRomanUpperCase(number).toLowerCase();
    }

    static toRomanUpperCase(number: number): string {
        if (number < 1 || number > 3999) return "";
        const thousands = ["", "M", "MM", "MMM"];
        const hundreds = ["", "C", "CC", "CCC", "CD", "D", "DC", "DCC", "DCCC", "CM"];
        const tens = ["", "X", "XX", "XXX", "XL", "L", "LX", "LXX", "LXXX", "XC"];
        const ones = ["", "I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX"];

        const idxThousands = Math.floor(number / 1000);
        const idxHundreds = Math.floor((number % 1000) / 100);
        const idxTens = Math.floor((number % 100) / 10);
        const idxOnes = number % 10;

        return thousands[idxThousands] + hundreds[idxHundreds] + tens[idxTens] + ones[idxOnes];
    }

    static toLatinAlphabetNumber(number: number, upperCase: boolean): string {
        return upperCase ? this.toLatinAlphabetNumberUpperCase(number) : this.toLatinAlphabetNumberLowerCase(number);
    }

    static toLatinAlphabetNumberLowerCase(number: number): string {
        return this.toLatinAlphabetNumberUpperCase(number).toLowerCase();
    }

    static toLatinAlphabetNumberUpperCase(number: number): string {
        if (number <= 0) return "";
        let sb = "";
        let n = number;
        while (n > 0) {
            n--;
            const charCode = 'A'.charCodeAt(0) + (n % 26);
            sb = String.fromCharCode(charCode) + sb;
            n = Math.floor(n / 26);
        }
        return sb;
    }

    static formatPageNumber(style: PageLabelNumberingStyle, number: number, prefix: string | null): string {
        let label = "";
        switch (style) {
            case PageLabelNumberingStyle.NONE:
                label = "";
                break;
            case PageLabelNumberingStyle.DECIMAL_ARABIC_NUMERALS:
                label = String(number);
                break;
            case PageLabelNumberingStyle.UPPERCASE_ROMAN_NUMERALS:
                label = this.toRomanUpperCase(number);
                break;
            case PageLabelNumberingStyle.LOWERCASE_ROMAN_NUMERALS:
                label = this.toRomanLowerCase(number);
                break;
            case PageLabelNumberingStyle.UPPERCASE_LETTERS:
                label = this.toLatinAlphabetNumberUpperCase(number);
                break;
            case PageLabelNumberingStyle.LOWERCASE_LETTERS:
                label = this.toLatinAlphabetNumberLowerCase(number);
                break;
        }

        return prefix ? prefix + label : label;
    }
}
